pkg/devfile/adapters/common/command.go

```diff
@@ -7,140 +7,156 @@ import (
 	"github.com/openshift/odo/pkg/devfile/parser/data"
 	"github.com/openshift/odo/pkg/devfile/parser/data/common"
 	"k8s.io/klog"
-
-	"github.com/pkg/errors"
 )
 
 // GetCommand iterates through the devfile commands and returns the associated devfile command
-func getCommand(data data.DevfileData, commandName string, required bool) (supportedCommand common.DevfileCommand, err error) {
-	for _, command := range data.GetCommands() {
-		if command.Name == commandName {
-
-			// Get the supported actions
-			supportedCommandActions, err := getSupportedCommandActions(data, command)
-
-			// None of the actions are supported so the command cannot be run
-			if len(supportedCommandActions) == 0 {
-				return supportedCommand, errors.Wrapf(err, "\nThe command \"%v\" was found but its actions are not supported", commandName)
-			} else if err != nil {
-				klog.Warning(errors.Wrapf(err, "The command \"%v\" was found but some of its actions are not supported", commandName))
+func getCommand(data data.DevfileData, commandName string, groupType common.DevfileCommandGroupType) (supportedCommand common.DevfileCommand, err error) {
+
+	commands := data.GetCommands()
+
+	for _, command := range commands {
+
+		command = updateGroupforCustomCommand(commandName, groupType, command)
+
+		// validate command
+		err = validateCommand(data, command)
+
+		if err != nil {
+			return common.DevfileCommand{}, err
+		}
+
+		// if command is specified via flags, it has the highest priority
+		// search through all commands to find the specified command name
+		// if not found fallback to error.
+		if commandName != "" {
+
+			if command.Exec.Id == commandName {
+
+				if command.Exec.Group.Kind == "" {
+					// Devfile V1 for commands passed from flags
+					// Group type is not updated during conversion
+					command.Exec.Group.Kind = groupType
+				}
+
+				// we have found the command with name, its groupType Should match to the flag
+				// e.g --build-command "mybuild"
+				// exec:
+				//   id: mybuild
+				// group:
+				//   kind: build
+				if command.Exec.Group.Kind != groupType {
+					return supportedCommand, fmt.Errorf("mismatched group kind, command %s is of group kind %v groupType in devfile", commandName, groupType)
```

Comment:
So the previous error output was like this: `The command \"%v\" was found but some of its actions are not supported` which was very useful to the user with regards to what went wrong. The output here is quite ambiguous. Can you elaborate more as to why the error is happening and why it's mismatched?

Reply:
@cdrage This is completely different error, let me explain it here
consider the below command specified in devfile
```
exec:
   id: "cmd-one"
   commandLine: "mvn compile"
   group:
        Kind: build
```
 now a user does odo push
`odo push --run-command cmd-one`
In this case it would error out, as command group specified in odo flag does not match specified in devfile

@maysunfaisal @cdrage  Would fix the strings in the err message and make it more intuitive to user.

Reply:
@maysunfaisal @cdrage regarding the formatting of error in message, it is present in v1 too, the source of error is not from here, it needs to be resolved with another PR..
```
$ odo push

Validation
 ✗  Validating the devfile [35201ns]
 ✗  Failed to start component with name nodejs.
Error: Failed to create the component: failed to validate devfile build and run commands: 
The command "devbuild" was found but its actions are not supported: Problem with command "devbuild" action #1: Actions must be of type "exec"%!!(MISSING)(EXTRA string=
)
```

Reply:
https://github.com/openshift/odo/issues/3280